### ISSUE TITLE
Application crash in async start method when completion handler is invoked twice

### ISSUE THREAD

**Issue by kairadiagne** [auth0/Auth0.swift#729]
### Checklist

- [X] The issue can be reproduced in the [Auth0.swift sample app](https://github.com/auth0-samples/auth0-ios-swift-sample/tree/master/Sample-01) (or N/A).
- [X] I have looked into the [README](https://github.com/auth0/Auth0.swift#readme) and have not found a suitable solution or answer.
- [X] I have looked into the [API documentation](https://auth0.github.io/Auth0.swift/) and have not found a suitable solution or answer.
- [X] I have searched the [issues](https://github.com/auth0/Auth0.swift/issues) and have not found a suitable solution or answer.
- [X] I have searched the [Auth0 Community](https://community.auth0.com/tags/c/sdks/5/swift) forums and have not found a suitable solution or answer.
- [X] I agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md).

### Description

Our QA team reported a crash in our login screen that can happen when we attempt to start the `ASWebAuthenticationSession` based authentication flow through the SDK. 

### What is the cause of the crash?
After closing the iOS control center, there is a short period where the scene's `UIWindow` is not in the active foreground state. Causing the `ASWebAuthenticationSession` to complete with the following error:

> ERROR: Error Domain=com.apple.AuthenticationServices.WebAuthenticationSession Code=3 "The UIWindowScene for the returned window was not in the foreground active state." UserInfo={NSDebugDescription=The UIWindowScene for the returned window was not in the foreground active state.} | The operation couldn’t be completed. (com.apple.AuthenticationServices.WebAuthenticationSession error 3.)

This was discussed in #446, and I agree that there is not much we can do to prevent this error. However, when using the async APIs, this can lead to the following crash:

> Thread 1: Fatal error: SWIFT TASK CONTINUATION MISUSE: start() tried to resume its continuation more than once, throwing userCancelled!

This means that the completion handler of the `start`  method [here](https://github.com/auth0/Auth0.swift/blob/9860ce65b9f9f361a5412d7ad3edb122198ba6ed/Auth0/Auth0WebAuth.swift#L297) is called twice. I didn't figure out why but It seems like that after the `ASWebAuthenticationSession` callback is called for the first error, the transaction store still has a reference to the transaction, which holds a reference to the callback. The next time we call the start method it calls the `store(_ transaction)` method which internally first cancels the current transaction. This causes the completion handler to be called twice.

### How to work around the issue?
We currently work around this issue by wrapping the start method in a continuation ourselves. We make the continuation optional and set it to nil after the completion block is called so that it can't be resumed again, like this:

``` Swift
  try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<SSOCredential, Swift.Error>) in
            var nillableContinuation: CheckedContinuation<SSOCredential, Swift.Error>? = continuation

            Auth0
                .webAuth()
                .useEphemeralSession()
                .start { result in
                    switch result {
                    case .success(let credential):
                        continuation.resume(with: .success(credential))
                        nillableContinuation = nil
                    case .failure(let error):
                        nillableContinuation?.resume(with: .failure(error))
                        nillableContinuation = nil
                    }
                }
        }
```

### Reproduction

1. Launch the app.
2. Now open/close the Control center when the app is on the login screen
3. After closing the control center, quickly tap the login button.
4. Observe that an error occurs.
5. Observe that tapping the button again crashes the app.

### Auth0.swift version

2.3.0

### Platform

iOS

### Platform version(s)

iOS 14, 15

### Xcode version

13.4.1

### Package manager

SPM

**Comment by Widcket** [comment]
Hi @kairadiagne, thanks for the detailed report. I'll take a look.

**Comment by josevazquez** [comment]
I have been seeing this exact behavior as well. I just tried out the `fix/async-await-race-condition`branch and can confirm that it fixed the `swift task continuation misuse` issue in my app.

**Comment by Widcket** [comment]
The fix is now out in [2.3.1](https://github.com/auth0/Auth0.swift/releases/tag/2.3.1). Thanks @kairadiagne for the high-quality bug report and @josevazquez for trying out the fix.

**Comment by kairadiagne** [comment]
@Widcket Thank you for spending time on this and finding a solution so quickly!
